src/sentry/utils/snowflake.py

```diff
@@ -0,0 +1,106 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import OrderedDict, Tuple
+
+from django.conf import settings
+from django.db import IntegrityError
+
+from sentry.utils import redis
+
+_TTL = timedelta(minutes=5)
+SENTRY_EPOCH_START = datetime(2022, 4, 26, 0, 0).timestamp()
+
+
+class SnowflakeIdMixin:
+    def save_with_snowflake_id(self, snowflake_redis_key, save_callback):
+        for _ in range(settings.MAX_REDIS_SNOWFLAKE_RETY_COUNTER):
+            if not self.id:
+                self.id = snowflake_id_generation(snowflake_redis_key)
+            try:
+                save_callback()
+                return
+            except IntegrityError:
+                self.id = None
+        raise Exception("Max allowed ID retry reached. Please try again in a second")
+
+
+@dataclass(frozen=True, eq=True)
+class SnowflakeBitSegment:
+    length: int
+    name: str
+
+    def validate(self, value):
+        if self.length <= 0:
+            raise Exception("The length should be a positive number")
+        if value >> self.length != 0:
+            raise Exception(f"{self.name} exceed max bit value of {self.length}")
+        return True
+
+
+ID_VALIDATOR = SnowflakeBitSegment(settings.SNOWFLAKE_ID_LENGTH, "Snowflake ID")
```

Comment:
I'd like to get rid of the `SNOWFLAKE_ID_LENGTH` constant (even if it's not a config value) and instead derive it from the other values. Something like

```python
SNOWFLAKE_ID_LENGTH = sum(segment.length for segment in BIT_SEGMENT_SCHEMA)
```

Since it's important to us that the length fit inside a JSON significand, you could even follow that up with

```python
assert SNOWFLAKE_ID_LENGTH == 53
```